src/Shared/PackageActions/NuGetPackageActions.cs

```diff
@@ -0,0 +1,169 @@
+        await using FileStream packageStream = new FileStream(
+            filePath,
+            FileMode.Create,
+            FileAccess.ReadWrite,
+            FileShare.ReadWrite,
+            bufferSize,
+            doExtract ? FileOptions.DeleteOnClose : FileOptions.None);
+        // If we want to extract the archive, delete the .nupgkg on close, otherwise keep it.
+
+        bool downloaded = await resource.CopyNupkgToStreamAsync(
+            packageUrl.Name,
+            NuGetVersion.Parse(packageUrl.Version),
+            packageStream,
+            _sourceCacheContext,
+            _logger, 
+            cancellationToken);
+        
+        // If the download failed, return null.
+        if (!downloaded)
+        {
+            return null;
+        }
+        
+        // If we want to extract the contents of the .nupkg, send it to ArchiveHelper.ExtractArchiveAsync.
+        if (doExtract)
+        {
+            return await ArchiveHelper.ExtractArchiveAsync(topLevelDirectory, Path.GetExtension(targetPath) + ".nupkg", packageStream, cached);
+        }
+
+        return filePath;
+    }
+
+    /// <inheritdoc />
+    public async Task<bool> DoesPackageExistAsync(
+        PackageURL packageUrl,
+        bool useCache = true,
+        CancellationToken cancellationToken = default)
+    {
+        FindPackageByIdResource resource = await _sourceRepository.GetResourceAsync<FindPackageByIdResource>();
+
+        // If a version is provided, check for existence of the specified version.
+        if (!string.IsNullOrWhiteSpace(packageUrl.Version))
+        {
+            bool exists = await resource.DoesPackageExistAsync(
+                packageUrl.Name,
+                NuGetVersion.Parse(packageUrl.Version),
+                _sourceCacheContext,
+                _logger, 
+                cancellationToken);
+
+            return exists;
+        }
+        
+        // If no version is provided, check to see if any versions exist on a package with the given name.
+        IEnumerable<NuGetVersion> versions = await resource.GetAllVersionsAsync(
+            packageUrl.Name,
+            _sourceCacheContext,
+            _logger, 
+            cancellationToken);
+
+        return versions.Any();
+    }
+
+    /// <inheritdoc />
+    public async Task<IEnumerable<string>> GetAllVersionsAsync(
+        PackageURL packageUrl,
+        bool includePrerelease = true,
+        bool useCache = true,
+        CancellationToken cancellationToken = default)
+    {
+        FindPackageByIdResource resource = await _sourceRepository.GetResourceAsync<FindPackageByIdResource>();
+
+        IEnumerable<NuGetVersion> versions = await resource.GetAllVersionsAsync(
+            packageUrl.Name,
+            _sourceCacheContext,
+            _logger, 
+            cancellationToken);
+
+        return versions
+            .Where(v => includePrerelease || !v.IsPrerelease)
+            .Select(v => v.ToString()).Reverse();
```

Comment:
Tiny nit: when working with fluent APIs like IEnumerable methods, I find it a bit easier to read when all of the fluent method calls have the same indent level. In this case, the `.Reverse()` to the right of the `Select` did not jump out to me until I thought more carefully about the sort order contract for this API: 
```suggestion
        return versions
            .Where(v => includePrerelease || !v.IsPrerelease)
            .Select(v => v.ToString())
            .Reverse();
```